pkg/dockerregistry/client.go

```diff
@@ -32,29 +35,73 @@ type Connection interface {
 
 // NewClient returns a client object which allows public access to
 // a Docker registry.
+// TODO: accept a docker auth config
 func NewClient() Client {
 	return &client{
-		connections: make(map[string]connection),
+		connections: make(map[string]*connection),
 	}
 }
 
 // client implements the Client interface
 type client struct {
-	connections map[string]connection
+	connections map[string]*connection
 }
 
+// Connect accepts the name of a registry in the common form Docker provides and will
+// create a connection to the registry. Callers may provide a host, a host:port, or
+// a fully qualified URL. When not providing a URL, the default scheme will be "https"
 func (c *client) Connect(name string) (Connection, error) {
-	if len(name) == 0 {
-		name = "index.docker.io"
+	target, err := normalizeRegistryName(name)
+	if err != nil {
+		return nil, err
 	}
-	if conn, ok := c.connections[name]; ok {
+	prefix := target.String()
+	if conn, ok := c.connections[prefix]; ok {
 		return conn, nil
 	}
-	conn := newConnection(name)
-	c.connections[name] = conn
+	conn := newConnection(*target)
+	c.connections[prefix] = conn
 	return conn, nil
 }
 
+func normalizeRegistryName(name string) (*url.URL, error) {
+	prefix := name
+	if len(prefix) == 0 {
+		prefix = "index.docker.io"
+	}
+	hadPrefix := false
+	switch {
+	case strings.HasPrefix(prefix, "http://"), strings.HasPrefix(prefix, "https://"):
```

Comment:
I'm not sure you can ever create an ImageStream with spec.dockerImageRepository set to http://... or https://....

```
$ cat v1-ncdc-busybox.json
{
  "apiVersion": "v1beta1",
  "kind": "ImageStream",
  "metadata": {
    "name": "v1ncdcbusybox"
  },
  "spec": {
    "dockerImageRepository": "http://localhost:5100/ncdc/busybox"
  }
}

$ osc create -f v1-ncdc-busybox.json
Error: ImageStream "v1ncdcbusybox" is invalid: spec.dockerImageRepository: invalid value 'http://localhost:5100/ncdc/busybox': the docker pull spec "http://localhost:5100/ncdc/busybox" must be two or three segments separated by slashes
```